scapy/contrib/automotive/enumerator.py

```diff
@@ -0,0 +1,316 @@
+                                 shortest_paths if node not in visited}
+            if not next_destinations:
+                return None
+            # next node is the destination with the lowest weight
+            current_node = min(next_destinations,
+                               key=lambda k: next_destinations[k][1])
+
+        # Work back through destinations in shortest path
+        path = []
+        while current_node is not None:
+            path.append(current_node)
+            next_node = shortest_paths[current_node][0]
+            current_node = next_node
+        # Reverse path
+        path = path[::-1]
+        return path
+
+
+class Enumerator(object):
+    """ Base class for Enumerators
+
+    Args:
+        sock: socket where enumeration takes place
+    """
+    description = "About my results"
+    negative_response_blacklist = []
+    ScanResult = namedtuple("ScanResult", "state req resp")
+
+    def __init__(self, sock):
+        self.sock = sock
+        self.results = list()
+        self.stats = {"answered": 0, "unanswered": 0, "answertime_max": 0,
+                      "answertime_min": 0, "answertime_avg": 0,
+                      "negative_resps": 0}
+        self.state_completed = defaultdict(bool)
+        self.last_state = None
+
+    @property
+    def completed(self):
+        return all([self.state_completed[s] for s in self.scanned_states])
+
+    def pre_scan(self, global_configuration):
+        pass
+
+    def scan(self, state, requests, **kwargs):
+        raise NotImplementedError
+
+    def post_scan(self, global_configuration):
+        pass
+
+    def update_stats(self):
+        answered = self.filtered_results
+        unanswered = [r for r in self.results if r.resp is None]
+        answertimes = [x.resp.time - x.req.sent_time for x in answered]
+        nrs = [r.resp for r in self.filtered_results if r.resp.service == 0x7f]
+        try:
+            self.stats["answered"] = len(answered)
+            self.stats["unanswered"] = len(unanswered)
+            self.stats["answertime_max"] = max(answertimes)
+            self.stats["answertime_min"] = min(answertimes)
+            self.stats["answertime_avg"] = mean(answertimes)
+            self.stats["negative_resps"] = len(nrs)
+        except ValueError:
+            for k, v in self.stats.items():
+                if v is None:
+                    self.stats[k] = 0
+
+    @property
+    def filtered_results(self):
+        return [r for r in self.results if r.resp is not None]
+
+    @property
+    def scanned_states(self):
+        return set([s for s, _, _, in self.results])
+
+    def show_negative_response_details(self, dump=False):
+        raise NotImplementedError("This needs a protocol specific "
+                                  "implementation")
+
+    def show(self, dump=False, filtered=True, verbose=False):
+        s = "\r\n\r\n" + "=" * (len(self.description) + 10) + "\n"
```

Comment:
You should only use `\n` here. For Linux it's fine then and for Windows python converts if it's ever written to a file:

```
PS C:\Users\Andreas\Desktop>python2
Python 2.7.17 (v2.7.17:c2f86d86e6, Oct 19 2019, 21:01:17) [MSC v.1500 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> f = open('test2.tmp', 'w')
>>> f.write('\n\n')
>>> f.close()
>>> exit()

PS C:\Users\Andreas\Desktop> Format-Hex test2.tmp

00000000   0D 0A 0D 0A
```

Reply:
Thanks